p42/src/song.rs

```diff
@@ -0,0 +1,215 @@
+    ///
+    /// let new_iter = SongIter { day: 1, index: 0 };
+    /// let s = new_iter.get_line();
+    /// assert_eq!(s.as_str(), "On the first day of Christmas,");
+    /// ```
+    pub fn get_line(&self) -> String {
+        match self.index {
+            0 => {
+                let mut first = String::from(LEFT);
+                first.push_str(DAYS[self.day - 1]);
+                first.push_str(RIGHT);
+                first
+            }
+            1 => String::from(SECOND),
+            i => {
+                if i == self.day + 2 {
+                    String::from("")
+                } else if i == self.day + 1 {
+                    let mut v = match self.day {
+                        1 => String::from(V1),
+                        _ => String::from(V2),
+                    };
+                    v.push_str(GIFTS[0]);
+                    match self.day {
+                        12 => v.push('!'),
+                        _ => v.push('.'),
+                    };
+                    v
+                } else {
+                    String::from(GIFTS[self.day - i + 1])
+                }
+            }
+        }
+    }
+
+    /// go to next state of iterator
+    pub fn increment(&mut self) {
+        self.index += 1;
+        if self.index > self.day + 2 {
+            self.day += 1;
+            self.index = 0;
+        }
+    }
+}
+
+impl Iterator for SongIter {
+    type Item = String;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.day > 12 || self.index > 13 {
+            None
+        } else {
+            let line = self.get_line();
+            self.increment();
+            Some(line)
+        }
+    }
+}
+
+/// add line number to each line in lyrics
+pub fn iter_with_line_num() -> impl Iterator<Item = String> {
+    let song_iter: SongIter = SongIter::default();
+    song_iter
+        .enumerate()
+        .map(|(num, line)| (num + 1).to_string() + ": " + &line)
+}
+
+/// extract lyrics from song iterator
+/// ```
+/// use p42::song::SongIter;
+/// use p42::song::song_to_string;
+/// use p42::song::iter_with_line_num;
+///
+/// assert_eq!(song_to_string(SongIter::default()), include_str!("lyrics_for_test.txt").replace("\r", ""));
+/// ```
+pub fn song_to_string(mut iter: impl Iterator<Item = String>) -> String {
+    let mut lyrics = iter.next().unwrap();
+    for line in iter {
+        lyrics.push('\n');
+        lyrics.push_str(line.as_str());
+    }
+    lyrics
+}
+
+/// extract lyrics from song iterator into file using song_to_iter
+pub fn song_to_file_1(iter: impl Iterator<Item = String>, path: &String) -> std::io::Result<()> {
+    std::fs::write(path, song_to_string(iter))?;
+    Ok(())
+}
+
+/// extract lyrics from song iterator into file without using song_to_iter
+pub fn song_to_file_2(
+    mut iter: impl Iterator<Item = String>,
+    path: &String,
+) -> std::io::Result<()> {
+    let mut file = std::io::LineWriter::new(std::fs::File::create(path)?);
```

Comment:
`LineWriter` is not necessary here, you could use the raw `File` or `BufWriter<File>`.